tests/scenario_tests/test_function_block_action.py

```diff
@@ -0,0 +1,213 @@
+import json
+import time
+
+from slack_sdk.signature import SignatureVerifier
+from slack_sdk.web import WebClient
+
+from slack_bolt.app import App
+from slack_bolt.request import BoltRequest
+from slack_bolt.context import BoltContext
+from slack_bolt.slack_function import SlackFunction
+
+from tests.mock_web_api_server import (
+    setup_mock_web_api_server,
+    cleanup_mock_web_api_server,
+    assert_auth_test_count,
+)
+from tests.utils import remove_os_env_temporarily, restore_os_env
+
+
+class TestFunctionBlockActions:
+    signing_secret = "secret"
+    valid_token = "xoxb-valid"
+    mock_api_server_base_url = "http://localhost:8888"
+    signature_verifier = SignatureVerifier(signing_secret)
+    web_client = WebClient(
+        token=valid_token,
+        base_url=mock_api_server_base_url,
+    )
+
+    def setup_method(self):
+        self.old_os_env = remove_os_env_temporarily()
+        setup_mock_web_api_server(self)
+
+    def teardown_method(self):
+        cleanup_mock_web_api_server(self)
+        restore_os_env(self.old_os_env)
+
+    def generate_signature(self, body: str, timestamp: str):
+        return self.signature_verifier.generate_signature(
+            body=body,
+            timestamp=timestamp,
+        )
+
+    def build_headers(self, timestamp: str, body: str):
+        return {
+            "content-type": ["application/json"],
+            "x-slack-signature": [self.generate_signature(body, timestamp)],
+            "x-slack-request-timestamp": [timestamp],
+        }
+
+    def build_request_from_body(self, message_body: dict) -> BoltRequest:
+        timestamp, body = str(int(time.time())), json.dumps(message_body)
+        return BoltRequest(body=body, headers=self.build_headers(timestamp, body))
+
+    def test_success(self):
```

Comment:
Can you add an extra test that verifies the decorator use cases? In my understanding, the following code should work too. If it does not work for some reason, I can explore possible ways too.
```python
func = app.function("c")
@func 
def main_func():
    pass

@func.action("a")
def handle_actions():
    pass
```

Reply:
@seratch I've added the test but it does not work exactly like the mentioned use case, here is the use case that currently works
```python
func = app.function("c")
@func 
def main_func():
    pass

@main_func.action("a")
def handle_actions():
    pass
```

This is the behavior I expected since `app.function` returns a `function` that then returns a `SlackFunction object` when called

From what I understand in order to implement the above use case the `SlackFunction object` would need to be returned by `app.function` and return itself when `SlackFunction object.__call__()` is used, this might go against the behavior of the other app decorators that return the `registered listener`.

Currently `SlackFunction object.__call__()` executes the `registered listener` and returns the output, `SlackFunction` tries to wrap the registered listener with an object that has additional decorators and  stores the callback_id.

Reply:
@WilliamBergamin Using the `main_func` method as a decorator object should be quite surprising for developers. I understand that the current implementation can work in this way, but I don't think we should go with this.